// plausible-tracker/build/main/lib/request.spec.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/* eslint-disable functional/no-let */
/* eslint-disable functional/immutable-data */
const request_1 = require("./request");
const getXhrMockClass = () => ({
    open: jest.fn(),
    send: jest.fn(),
    setRequestHeader: jest.fn(),
    onreadystatechange: jest.fn(),
    readyState: 0,
    ready(state) {
        this.readyState = state;
        this.onreadystatechange();
    },
});
let xhrMockClass;
const xmr = jest.spyOn(window, 'XMLHttpRequest');
const defaultData = {
    hashMode: false,
    trackLocalhost: false,
    url: 'https://my-app.com/my-url',
    domain: 'my-app.com',
    referrer: null,
    deviceWidth: 1080,
    apiHost: 'https://plausible.io',
};
const oldLocation = window.location;
const windowLocation = JSON.stringify(window.location);
// eslint-disable-next-line @typescript-eslint/ban-ts-comment
// @ts-ignore
delete window['location'];
Object.defineProperty(window, 'location', {
    value: JSON.parse(windowLocation),
});
beforeEach(() => {
    jest.clearAllMocks();
    xhrMockClass = getXhrMockClass();
    xmr.mockReturnValue(xhrMockClass);
    window.location.hostname = 'my-app.com';
    window.location.protocol = 'http:';
});
afterAll(() => {
    Object.defineProperty(window, 'location', oldLocation);
});
describe('sendEvent', () => {
    test('sends default event', () => {
        expect(xmr).not.toHaveBeenCalled();
        request_1.sendEvent('myEvent', defaultData);
        xhrMockClass.ready(4);
        expect(xmr).toHaveBeenCalledTimes(1);
        expect(xhrMockClass.open).toHaveBeenCalledWith('POST', `${defaultData.apiHost}/api/event`, true);
        expect(xhrMockClass.setRequestHeader).toHaveBeenCalledWith('Content-Type', 'text/plain');
        const payload = {
            n: 'myEvent',
            u: defaultData.url,
            d: defaultData.domain,
            r: defaultData.referrer,
            w: defaultData.deviceWidth,
            h: 0,
        };
        expect(xhrMockClass.send).toHaveBeenCalledWith(JSON.stringify(payload));
    });
    test('hash mode', () => {
        expect(xmr).not.toHaveBeenCalled();
        request_1.sendEvent('myEvent', Object.assign(Object.assign({}, defaultData), { hashMode: true }));
        const payload = {
            n: 'myEvent',
            u: defaultData.url,
            d: defaultData.domain,
            r: defaultData.referrer,
            w: defaultData.deviceWidth,
            h: 1,
        };
        expect(xhrMockClass.send).toHaveBeenCalledWith(JSON.stringify(payload));
    });
    test('does not send to localhost', () => {
        window.location.hostname = 'localhost';
        expect(xmr).not.toHaveBeenCalled();
        request_1.sendEvent('myEvent', defaultData);
        expect(xmr).not.toHaveBeenCalled();
    });
    test('does not send if local file', () => {
        window.location.protocol = 'file:';
        expect(xmr).not.toHaveBeenCalled();
        request_1.sendEvent('myEvent', defaultData);
        expect(xmr).not.toHaveBeenCalled();
    });
    test('sends to localhost if trackLocalhost', () => {
        window.location.hostname = 'localhost';
        expect(xmr).not.toHaveBeenCalled();
        request_1.sendEvent('myEvent', Object.assign(Object.assign({}, defaultData), { trackLocalhost: true }));
        expect(xmr).toHaveBeenCalled();
    });
    test('does not send if "plausible_ignore" is set to "true" in localStorage', () => {
        window.localStorage.setItem('plausible_ignore', 'true');
        expect(xmr).not.toHaveBeenCalled();
        request_1.sendEvent('myEvent', defaultData);
        expect(xmr).not.toHaveBeenCalled();
        window.localStorage.setItem('plausible_ignore', 'something-not-true');
        request_1.sendEvent('myEvent', defaultData);
        expect(xmr).toHaveBeenCalled();
        window.localStorage.removeItem('plausible_ignore');
    });
    test('does not throw an error if localStorage is not available', () => {
        const localStorage = window.localStorage;
        // eslint-disable-next-line @typescript-eslint/ban-ts-comment
        // @ts-ignore
        delete window['localStorage'];
        expect(xmr).not.toHaveBeenCalled();
        request_1.sendEvent('myEvent', defaultData);
        expect(xmr).toHaveBeenCalled();
        // eslint-disable-next-line @typescript-eslint/ban-ts-comment
        // @ts-ignore
        window.localStorage = localStorage;
    });
    test('calls callback', () => {
        expect(xmr).not.toHaveBeenCalled();
        const callback = jest.fn();
        request_1.sendEvent('myEvent', defaultData, { callback });
        expect(callback).not.toHaveBeenCalled();
        xhrMockClass.ready(3);
        expect(callback).not.toHaveBeenCalled();
        xhrMockClass.ready(4);
        expect(callback).toHaveBeenCalled();
    });
    test('sends props', () => {
        expect(xmr).not.toHaveBeenCalled();
        const props = { variation1: 'A', variation2: 'B' };
        request_1.sendEvent('myEvent', defaultData, { props });
        const payload = {
            n: 'myEvent',
            u: defaultData.url,
            d: defaultData.domain,
            r: defaultData.referrer,
            w: defaultData.deviceWidth,
            h: 0,
            p: JSON.stringify(props),
        };
        expect(xhrMockClass.send).toHaveBeenCalledWith(JSON.stringify(payload));
    });
});
