// plausible-tracker/build/main/lib/tracker.spec.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
/* eslint-disable functional/immutable-data */
const requestModule = __importStar(require("./request"));
const tracker_1 = __importDefault(require("./tracker"));
const requestSpy = jest.spyOn(requestModule, 'sendEvent');
beforeEach(() => {
    jest.clearAllMocks();
    requestSpy.mockImplementation();
});
describe('tracker', () => {
    const getDefaultData = () => ({
        hashMode: false,
        trackLocalhost: false,
        url: location.href,
        domain: location.hostname,
        referrer: document.referrer || null,
        deviceWidth: window.innerWidth,
        apiHost: 'https://plausible.io',
    });
    const getCustomData = () => ({
        hashMode: true,
        trackLocalhost: true,
        url: 'https://my-url.com',
        domain: 'my-domain.com',
        referrer: 'my-referrer',
        deviceWidth: 1080,
        apiHost: 'https://my-api-hos.t',
    });
    const getEventOptions = () => ({
        callback: jest.fn(),
        props: {
            variation1: 'A',
            variation2: 'B',
            variation3: 1,
            variation4: true,
        },
    });
    test('inits with default config', () => {
        const { trackEvent } = tracker_1.default();
        expect(requestSpy).not.toHaveBeenCalled();
        trackEvent('myEvent');
        expect(requestSpy).toHaveBeenCalled();
        expect(requestSpy).toHaveBeenCalledWith('myEvent', expect.objectContaining(getDefaultData()), undefined);
    });
    test('inits with overridden config', () => {
        const config = getCustomData();
        const { trackEvent } = tracker_1.default(config);
        expect(requestSpy).not.toHaveBeenCalled();
        trackEvent('myEvent');
        expect(requestSpy).toHaveBeenCalled();
        expect(requestSpy).toHaveBeenCalledWith('myEvent', config, undefined);
    });
    describe('trackEvent', () => {
        test('tracks event', () => {
            const { trackEvent } = tracker_1.default();
            expect(requestSpy).not.toHaveBeenCalled();
            trackEvent('myEvent');
            expect(requestSpy).toHaveBeenCalled();
            expect(requestSpy).toHaveBeenCalledWith('myEvent', getDefaultData(), undefined);
        });
        test('accepts data on trackEvent', () => {
            const { trackEvent } = tracker_1.default();
            expect(requestSpy).not.toHaveBeenCalled();
            const config = getCustomData();
            trackEvent('myEvent', undefined, config);
            expect(requestSpy).toHaveBeenCalled();
            expect(requestSpy).toHaveBeenCalledWith('myEvent', config, undefined);
        });
        test('accepts options on trackEvent', () => {
            const { trackEvent } = tracker_1.default();
            expect(requestSpy).not.toHaveBeenCalled();
            const options = getEventOptions();
            trackEvent('myEvent', options);
            expect(requestSpy).toHaveBeenCalled();
            expect(requestSpy).toHaveBeenCalledWith('myEvent', getDefaultData(), options);
        });
    });
    describe('pageView', () => {
        test('tracks pageview', () => {
            const { trackPageview } = tracker_1.default();
            expect(requestSpy).not.toHaveBeenCalled();
            trackPageview();
            expect(requestSpy).toHaveBeenCalled();
            expect(requestSpy).toHaveBeenCalledWith('pageview', getDefaultData(), undefined);
        });
        test('accepts data on pageview', () => {
            const { trackPageview } = tracker_1.default();
            expect(requestSpy).not.toHaveBeenCalled();
            const config = getCustomData();
            trackPageview(config);
            expect(requestSpy).toHaveBeenCalled();
            expect(requestSpy).toHaveBeenCalledWith('pageview', config, undefined);
        });
        test('accepts event options on pageview', () => {
            const { trackPageview } = tracker_1.default();
            expect(requestSpy).not.toHaveBeenCalled();
            const options = getEventOptions();
            trackPageview({}, options);
            expect(requestSpy).toHaveBeenCalled();
            expect(requestSpy).toHaveBeenCalledWith('pageview', getDefaultData(), options);
        });
    });
    describe('enableAutoPageviews', () => {
        test('tracks first pageview', () => {
            const { enableAutoPageviews } = tracker_1.default();
            expect(requestSpy).not.toHaveBeenCalled();
            const cleanup = enableAutoPageviews();
            expect(requestSpy).toHaveBeenCalledWith('pageview', getDefaultData(), undefined);
            cleanup();
        });
        test('does not track popstate if no history.pushState', () => {
            const originalPushState = history.pushState;
            const { enableAutoPageviews } = tracker_1.default();
            expect(requestSpy).not.toHaveBeenCalled();
            history.pushState = null;
            const cleanup = enableAutoPageviews();
            expect(requestSpy).toHaveBeenCalledTimes(1);
            window.dispatchEvent(new PopStateEvent('popstate'));
            expect(requestSpy).toHaveBeenCalledTimes(1);
            cleanup();
            history.pushState = originalPushState;
        });
        test('tracks pageviews on push state', () => {
            const { enableAutoPageviews } = tracker_1.default();
            expect(requestSpy).not.toHaveBeenCalled();
            const cleanup = enableAutoPageviews();
            expect(requestSpy).toHaveBeenCalledTimes(1);
            history.pushState({}, 'second', '/url');
            expect(requestSpy).toHaveBeenCalledTimes(2);
            cleanup();
        });
        test('tracks pageviews on popstate', () => {
            const { enableAutoPageviews } = tracker_1.default();
            expect(requestSpy).not.toHaveBeenCalled();
            const cleanup = enableAutoPageviews();
            expect(requestSpy).toHaveBeenCalledTimes(1);
            window.dispatchEvent(new PopStateEvent('popstate'));
            expect(requestSpy).toHaveBeenCalledTimes(2);
            cleanup();
        });
        test('does not track hashchange by default', () => {
            const { enableAutoPageviews } = tracker_1.default();
            expect(requestSpy).not.toHaveBeenCalled();
            const cleanup = enableAutoPageviews();
            expect(requestSpy).toHaveBeenCalledTimes(1);
            window.dispatchEvent(new HashChangeEvent('hashchange'));
            expect(requestSpy).toHaveBeenCalledTimes(1);
            cleanup();
        });
        test('tracks hashchange if specified', () => {
            const { enableAutoPageviews } = tracker_1.default({ hashMode: true });
            expect(requestSpy).not.toHaveBeenCalled();
            const cleanup = enableAutoPageviews();
            expect(requestSpy).toHaveBeenCalledTimes(1);
            window.dispatchEvent(new HashChangeEvent('hashchange'));
            expect(requestSpy).toHaveBeenCalledTimes(2);
            cleanup();
        });
    });
    describe('enableAutoOutboundTracking', () => {
        const createAnchors = (amount, start = 0) => {
            return [...Array(amount).keys()].map((i) => {
                const el = document.createElement('a');
                el.setAttribute('href', `https://link-${i + start}.com`);
                return el;
            });
        };
        const mountAnchors = (anchors) => {
            const div = document.createElement('div');
            anchors.forEach((a) => div.appendChild(a));
            document.body.append(div);
        };
        const clickLink = (anchor) => {
            anchor.click();
        };
        afterEach(() => {
            // Reset body
            document.body.innerHTML = '';
        });
        test('tracks existing anchors', () => {
            const anchors = createAnchors(10);
            mountAnchors(anchors);
            const { enableAutoOutboundTracking } = tracker_1.default(document);
            const cleanup = enableAutoOutboundTracking();
            expect(requestSpy).not.toHaveBeenCalled();
            // Click anchors
            anchors.forEach(clickLink);
            expect(requestSpy).toHaveBeenCalledTimes(10);
            anchors.forEach((a) => {
                expect(requestSpy).toHaveBeenCalledWith('Outbound Link: Click', expect.anything(), expect.objectContaining({
                    props: { url: a.href },
                }));
            });
            cleanup();
        });
        test('do not track host links', () => {
            const anchors = createAnchors(1);
            mountAnchors(anchors);
            const anchor = anchors[0];
            anchor.setAttribute('href', `https://${location.host}`);
            const { enableAutoOutboundTracking } = tracker_1.default(document);
            const cleanup = enableAutoOutboundTracking();
            expect(requestSpy).not.toHaveBeenCalled();
            // Click anchors
            anchors.forEach(clickLink);
            expect(requestSpy).not.toHaveBeenCalled();
            cleanup();
        });
        test('tracks href change', async () => {
            const anchors = createAnchors(1);
            mountAnchors(anchors);
            const { enableAutoOutboundTracking } = tracker_1.default(document);
            const cleanup = enableAutoOutboundTracking();
            const newHref = 'https://other.com/';
            // Change href
            const anchor = document.querySelector('a');
            expect(anchor).toBeTruthy();
            if (anchor) {
                anchor.removeAttribute('href');
                anchor.setAttribute('href', newHref);
            }
            expect(requestSpy).not.toHaveBeenCalled();
            // Wait for observer to run
            await new Promise((resolve) => setTimeout(() => {
                resolve(null);
            }, 100));
            // Click anchors
            anchors.forEach(clickLink);
            expect(requestSpy).toHaveBeenCalledTimes(1);
            expect(requestSpy).toBeCalledWith('Outbound Link: Click', expect.anything(), expect.objectContaining({
                props: { url: newHref },
            }));
            cleanup();
        });
        test('tracks node addition', async () => {
            const oldAnchors = createAnchors(5);
            mountAnchors(oldAnchors);
            const { enableAutoOutboundTracking } = tracker_1.default(document);
            const cleanup = enableAutoOutboundTracking();
            const newAnchors = createAnchors(5, 5);
            mountAnchors(newAnchors);
            expect(requestSpy).not.toHaveBeenCalled();
            const anchors = [...oldAnchors, ...newAnchors];
            // Wait for observer to run
            await new Promise((resolve) => setTimeout(() => {
                resolve(null);
            }, 100));
            // Click anchors
            anchors.forEach((a) => a.click());
            expect(requestSpy).toHaveBeenCalledTimes(10);
            anchors.forEach((a) => {
                expect(requestSpy).toHaveBeenCalledWith('Outbound Link: Click', expect.anything(), expect.objectContaining({
                    props: { url: a.href },
                }));
            });
            cleanup();
        });
        test('tracks node removal', async () => {
            var _a;
            const anchors = createAnchors(5);
            mountAnchors(anchors);
            const { enableAutoOutboundTracking } = tracker_1.default(document);
            const cleanup = enableAutoOutboundTracking();
            expect(requestSpy).not.toHaveBeenCalled();
            // Remove nodes
            (_a = document.querySelector('div')) === null || _a === void 0 ? void 0 : _a.remove();
            // Wait for observer to run
            await new Promise((resolve) => setTimeout(() => {
                resolve(null);
            }, 100));
            // Click anchors
            anchors.forEach(clickLink);
            expect(requestSpy).not.toHaveBeenCalled();
            cleanup();
        });
    });
});
