// react-scroll-to-bottom/lib/ScrollToBottom/AutoHideFollowButton.js
"use strict";

var _Object$defineProperty = require("@babel/runtime-corejs3/core-js-stable/object/define-property");

var _interopRequireDefault = require("@babel/runtime-corejs3/helpers/interopRequireDefault");

_Object$defineProperty(exports, "__esModule", {
  value: true
});

exports["default"] = void 0;

var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime-corejs3/helpers/slicedToArray"));

var _classnames = _interopRequireDefault(require("classnames"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireDefault(require("react"));

var _useScrollToEnd = _interopRequireDefault(require("../hooks/useScrollToEnd"));

var _useSticky3 = _interopRequireDefault(require("../hooks/useSticky"));

var _useStyleToClassName = _interopRequireDefault(require("../hooks/internal/useStyleToClassName"));

var ROOT_STYLE = {
  backgroundColor: 'rgba(0, 0, 0, .2)',
  borderRadius: 10,
  borderWidth: 0,
  bottom: 5,
  cursor: 'pointer',
  height: 20,
  outline: 0,
  position: 'absolute',
  right: 20,
  width: 20,
  '&:hover': {
    backgroundColor: 'rgba(0, 0, 0, .4)'
  },
  '&:active': {
    backgroundColor: 'rgba(0, 0, 0, .6)'
  }
};

var AutoHideFollowButton = function AutoHideFollowButton(_ref) {
  var children = _ref.children,
      className = _ref.className;

  var _useSticky = (0, _useSticky3["default"])(),
      _useSticky2 = (0, _slicedToArray2["default"])(_useSticky, 1),
      sticky = _useSticky2[0];

  var rootCSS = (0, _useStyleToClassName["default"])()(ROOT_STYLE);
  var scrollToEnd = (0, _useScrollToEnd["default"])();
  return !sticky && /*#__PURE__*/_react["default"].createElement("button", {
    className: (0, _classnames["default"])(rootCSS, (className || '') + ''),
    onClick: scrollToEnd,
    type: "button"
  }, children);
};

AutoHideFollowButton.defaultProps = {
  children: undefined,
  className: ''
};
AutoHideFollowButton.propTypes = {
  children: _propTypes["default"].any,
  className: _propTypes["default"].string
};
var _default = AutoHideFollowButton;
exports["default"] = _default;
